### PR TITLE
gh-85984: Add POSIX pseudo-terminal functions.

### PR DESCRIPTION
**Pull request by 8vasu**
This follows https://github.com/python/cpython/pull/101831. This is one in a series of PRs aimed at cleaning-up, fixing bugs in, introducing new features in, and updating the code in "Lib/pty.py".

This PR answers the following question: `os.forkpty()` and `pty.fork()` return a pair `pid, fd`, where `fd` is a file descriptor of the master end of a pseudo-terminal pair; if at a later stage one needs to make some modifications to the slave end (such as setting termios attributes), then how does one obtain access to it without reimplementing `ptsname()` in Python or loading it from a shared library like someone is doing here: https://stackoverflow.com/questions/52338062/calling-libc-select-from-python-from-pty-master-side?

This is a dependency of https://github.com/python/cpython/pull/101833, which only needs `os.ptsname()`. However, this PR also adds `os.posix_openpt()`, `os.grantpt()`, and `os.unlockpt()` since all of these POSIX functions are "companions" of each other.

Signed-off-by: Soumendra Ganguly [soumendraganguly@gmail.com](mailto:soumendraganguly@gmail.com)

* Issue: gh-85984

### CODE REVIEW

**Comment by 8vasu** [comment]
@gpshead This PR introduces 4 new functions which are simple wrappers. The rationale is explained in the main comment, and other details including testing instructions are provided as inline comments on the diffs.

For now I have converted https://github.com/python/cpython/pull/101832 and https://github.com/python/cpython/pull/101833 to drafts since they require more work and are slightly more ambitious projects.

As always, no rush.

File: Doc/library/os.rst
Comment (8vasu) [inline Doc/library/os.rst:1117]: While I have used the term "system call" here and in other places for the other three functions, these are under section 3 ("library functions") as opposed to section 2 ("system calls") of Linux and most BSD manpages, except `posix_openpt()` is under section 2 of FreeBSD. Is there a better term that we should use?

File: Modules/posixmodule.c
Comment (8vasu) [inline Modules/posixmodule.c:7702]: The POSIX `grantpt()` manpage (https://pubs.opengroup.org/onlinepubs/9699919799/) says

> The behavior of the grantpt() function is unspecified if the application has installed a signal handler to catch SIGCHLD signals.

What we are doing here is exactly how the fallback implementation of `os.openpty()` makes sure that behavior is not unpredictable: a `PyOS_setsig` call replaces the custom `SIGCHLD` handler with the default handler followed by the `grantpt()` call followed by restoring the original `SIGCHLD` handler.

File: Modules/posixmodule.c
Comment (8vasu) [inline Modules/posixmodule.c:7751]: Note that

> import os
master, slave = os.openpty()
print(os.ttyname(slave))

and

> import os
master, slave = os.openpty()
print(os.ptsname(master))

will give same output, but `os.ptsname()` works when a file descriptor of the slave is not available!

File: Modules/posixmodule.c
Comment (8vasu) [inline Modules/posixmodule.c:8376]: **Test for os.ptsname()**

```
import os
master, slave = os.openpty()
if os.ptsname(master) == os.ttyname(slave):
      print("Success")
else:
      print("Failure")
```
---

**Test for the entire family**

To test these functions, mimic the small example program at the bottom of the POSIX manpage of `posix_openpt()` (https://pubs.opengroup.org/onlinepubs/9699919799/):

```
import os
master_fd = os.posix_openpt(os.O_RDWR|os.O_NOCTTY)
os.grantpt(master_fd)
os.unlockpt(master_fd)
slave_path = os.ptsname(master_fd)
print("slave path:", slave_path)
slave_fd = os.open(slave_path, os.O_RDWR|os.O_NOCTTY)
```
Note that the above is roughly how `openpty()` is implemented on many systems including CPython's fallback `os.openpty()` implementation. This should fail if the `os.unlockpt()` call is commented out. On my Debian 11 system, commenting out the `os.grantpt()` call does **not** result in failure, and this is the expected behavior of the  `grantpt()` C function on my system: please read the next section if you are more interested in this phenomenon.

---

**Note on grantpt() behavior**

From the Linux manpage (https://man7.org/linux/man-pages/man3/grantpt.3.html):

> Many systems implement this function via a set-user-ID helper
       binary called "pt_chown".  On Linux systems with a devpts
       filesystem (present since Linux 2.2), the kernel normally sets
       the correct ownership and permissions for the pseudoterminal
       slave when the master is opened ([posix_openpt(3)](https://man7.org/linux/man-pages/man3/posix_openpt.3.html)), so that
       nothing must be done by grantpt().  Thus, no such helper binary
       is required (and indeed it is configured to be absent during the
       glibc build that is typical on many systems).

To avoid security issues due to the presence of the setuid binary `pt_chown`, my glibc `grantpt()` does not do anything and instead depends on the kernel to set correct ownership and permission attributes of the slave device. Note that `grantpt()` is supposed to set the permissions of the slave to `0620 (crw--w----)`. To demonstrate that the default glibc `grantpt()` on Debian 11 indeed depends on the kernel and does not to anything by itself, try the following.

Let us assume that all steps are being executed by a non-root user. In an interactive Python session, run

```
>>> import os
>>> master_fd = os.posix_openpt(os.O_RDWR|os.O_NOCTTY)
>>> os.ptsname(master_fd)
```

Then go to a different OS shell session and run

```
$ chmod u-w <path-to-slave-from-os-ptsname>
```

Back in Python, run

```
>>> os.grantpt(master_fd)
```

Go back to the OS shell and run

```
$ ls -l <path-to-slave-from-os-ptsname>
```

to see that `os.grantpt()` was unable to change permissions to `0620 (crw--w----)` (but did not fail or raise exceptions).

**Comment by arhadthedev** [comment]
@gpshead as a reviewer of the preceeding gh-101831.

File: Modules/posixmodule.c
Comment (8vasu) [reply Modules/posixmodule.c:7751]: **Notes on thread safety**

Should we add something in the docs about thread safety of these functions?

---

The POSIX manpage (https://pubs.opengroup.org/onlinepubs/9699919799/) of `ptsname()` says the following:

> The ptsname() function need not be thread-safe.

FreeBSD's libc and glibc include a reentrant version called `ptsname_r()` corresponding to the pair `ttyname()`, `ttyname_r()`. However, the `ttyname*` pair is POSIX, whereas only `ptsname()` is POSIX. As a result, `ptsname_r()` has portability issues: https://www.gnu.org/software/gnulib/manual/html_node/ptsname_005fr.html. However, the Linux manpage (https://man7.org/linux/man-pages/man3/ptsname.3.html) claims the following:

> ptsname_r() is a Linux extension, that is proposed for inclusion in the next major revision of POSIX.1 (Issue 8).

If `ptsname_r()` becomes standardized, then we can replace our `ptsname()` call in `os.ptsname()` with `ptsname_r()` in the future. Note that the `os.ttyname()` implementation uses `ttyname_r()` instead of `ttyname()`.

---

The POSIX manpage of `ptsname()` also says the following:

> The application shall not modify the string returned. The returned pointer might be invalidated or the string content might be overwritten by a subsequent call to ptsname(). The returned pointer and the string content might also be invalidated if the calling thread is terminated.

In fact, the Linux manpage says the following:

> On success, ptsname() returns a pointer to a string in static storage which will be overwritten by subsequent calls.  This pointer must not be freed.  On failure, NULL is returned.

Therefore, (almost) simultaneous calls to `os.ptsname()` from different threads A and B such that B calls the system `ptsname()` right in between the `ptsname()` and `PyUnicode_DecodeFSDefault()` calls in A (below) is a problematic situation.

File: configure.ac
Comment (erlend-aasland) [inline configure.ac]: IIRC, AC_SEARCH_LIBS affects the global LIBS. We probably do not want that.

Comment (8vasu) [reply configure.ac]: Thank you for the review. I had added `os.login_tty()` in https://github.com/python/cpython/pull/29658, where I had originally used `AC_CHECK_FUNCS`. However, @tiran suggested that I use `AC_SEARCH_LIBS` instead, mentioning the following:

> This approach is easier and better. It updates LIBS only when it is necessary and won't add libutils a second time. The login_tty is provided by libutil on Linux, FreeBSD, OpenBSD, and NetBSD.

Should I still go ahead and change them to `AC_CHECK_FUNCS`? Or is there yet another approach that is better?

Edit: fixed grammar

Comment (erlend-aasland) [reply configure.ac]: You could wrap this whole section with `WITH_SAVE_ENV` just to be sure LIBS (or anything else) is touched. You can grep `configure.ac` for example use.

Comment (8vasu) [reply configure.ac]: Thank you for your guidance. I will do this now.

**Comment by 8vasu** [comment]
@erlend-aasland Thank you for being patient.

Turns out that unlike unlike `openpty()`, `login_tty()`, and `forkpty()`, these 4 POSIX functions are a part of the standard C library on *nix and not a part of `libutil`. Therefore, I just needed to add the function names in an `AC_CHECK_FUNCS` block in `configure.ac`.

However, maybe we need to make a separate PR to modify the `openpty()`, `login_tty()`, `forkpty()` block in `configure.ac` to use `WITH_SAVE_ENV`?

**Review by erlend-aasland — APPROVED** [review]

**Comment by 8vasu** [comment]
@erlend-aasland Thank you!

**Comment by 8vasu** [comment]
@gpshead Just a gentle reminder. No rush.

File: Modules/posixmodule.c
Comment (encukou) [reply Modules/posixmodule.c:8376]: Is it possible to automate these?

File: Modules/posixmodule.c
Comment (encukou) [inline Modules/posixmodule.c]: This relies on the GIL. IMO, that shouldn't block this PR for 3.13, but I'd like to follow up.

@colesbury, I plan to do lots of reviews this year. Is there anything I should do when I spot a case like this?

File: Doc/library/os.rst
Comment (encukou) [inline Doc/library/os.rst:1452]: Should the added `O_CLOEXEC` flag be mentioned here?

File: Modules/posixmodule.c
Comment (encukou) [reply Modules/posixmodule.c:8376]: (I can help, but I don't know if you've already tried.)

Comment (8vasu) [reply Modules/posixmodule.c:8376]: Thank you for the review! Apologies for not attending this earlier: I recently moved to a different country.

Correct me if I am wrong, but I do not think there is a file containing tests for `posixmodule.c`. Maybe just put these in a shell script for temporary testing? Do you want me to do this?

File: Modules/posixmodule.c
Comment (colesbury) [reply Modules/posixmodule.c]: Let's track these in an issue with the "topic-free-threading" label

File: Modules/posixmodule.c
Comment (8vasu) [reply Modules/posixmodule.c:8376]: Edit: Grammar: *... not attending TO this earlier ...

File: Doc/library/os.rst
Comment (8vasu) [reply Doc/library/os.rst:1452]: Looks like @gpshead added that to make the returned master fd non-inheritable. According to his commit:

> By default all fds returned by Python APIs must be non-inheritable.

Given this information, I will let you and @gpshead be the judge and I will make the changes accordingly.

File: Modules/posixmodule.c
Comment (8vasu) [reply Modules/posixmodule.c:8376]: I have added a comment about the `O_CLOEXEC` above. I do not remember if I performed these tests after @gpshead made his commits. Therefore, I think I will test again (maybe I will write a shell script to automate these).

Comment (encukou) [reply Modules/posixmodule.c:8376]: I was thinking of adding these tests to the CI test suite. I'll take a stab at it next week.

Comment (8vasu) [reply Modules/posixmodule.c:8376]: Thank you so much. Please keep me in the loop so that I can learn how to do this.

Comment (encukou) [reply Modules/posixmodule.c:8376]: I've added some tests as a commit to this PR. Do they look understandable? Is there anything you'd add?

File: Doc/library/os.rst
Comment (encukou) [reply Doc/library/os.rst:1452]: I'll gladly defer to Greg, but IMO this needs to be the default, and it's not necessary to provide a knob for it (yet?).
But the docs should mention the effect and implementation: the new file descriptor is [non-inheritable]; `O_CLOEXEC` flag is added to *flags*.

**Comment by encukou** [comment]
Reading the manpages further, I see that Linux has a reentrant variant of `ptsname`:

> ptsname_r() is a Linux extension, that is proposed for inclusion in the next major revision of POSIX.1 (Issue 8).

IMO, `os.ptsname` should call this on Linux if it's available (with a `char buffer[MAXPATHLEN+1]`), and the docs should note that the function isn't thread-safe on other systems.

**Comment by encukou** [comment]
(If you prefer to leave `ptsname_r` to another PR, that's an option too.)

**Comment by 8vasu** [comment]
@encukou I prefer to include `ptsname_r` in this PR; thank you for the suggestion and sorry about the late reply. This is on my TODO list; please give me 1 more day (or maybe 2).

**Comment by encukou** [comment]
There's no rush, take your time :)

**Comment by 8vasu** [comment]
@encukou 

I have made the changes you requested and more; please take a look at them when you get time:
1. Change `os.grantpt()` so that the correct `errno` is set before calling `posix_error()`.
2. Change `os.ptsname()` so that it uses `os.ptsname_r()` when available; reflect this in docs.
3. Mention `OS_CLOEXEC` in the documentation for `os.posix_openpt()`.
4. Mention in the docs that these functions do not close the master fd upon failure.
5. One of the tests that you added was using `os.openpty()` followed by `os.grantpt()` and `os.unlockpt()`. That did not make sense to me, since I think `openpty()` C implementations do this already? Please let me know if this is wrong. I added a new test that follows the standard sequence `posix_openpt() -> grantpt() -> unlockpt() -> ptsname()` that this new test subsumed test that you had written for `ptsname()`.
6. I used https://github.com/tiran/cpython_autoconf (autoconf version 2.71) written by @tiran to generate the configure script. The test `Tests / Check if generated files are up to date (pull_request)` failed; maybe I should use something else now? The default autoconf on my Debian system is also (GNU Autoconf) 2.71, so generating using that will not make a difference I think.
7. About the pty pair terminology: I noticed that you used the naming scheme `main, second`. I used the alternative `mother, son` as a test to see if you like it. While I am not a CPython maintainer and hence do not have much say in deciding which directions such a prestigious project should take, I have severe clinical OCD/OCPD and would like to issue a warning here: using generic names like `main_fd, second_fd`, or `parent_fd, child_fd`, or `server_fd, client_fd` **will** lead to havoc very easily. Just take a look at the source of your very own https://github.com/python/cpython/blob/3.12/Lib/pty.py and try a search/replace involving terms parent/child (already used by processes) or a program like `ssh(1)` which uses the terms server/client in networking context  and also sets up a pty pair. On the other hand one can easily have `first_fd, second_fd, third_fd` or even `main_fd` in their existing programs.  The pair `main, subordinate` comes to mind, but again, `main_fd` might be taken and `main, subordinate` carry the same overtones as `master, slave`, which is what you are trying to avoid in the first place. For the most comfortable change, we must use something that is -
   - not already taken to almost absolutely eliminate any chances of collision while
   - maintaining that the initials `m` and `s` stay the same.

**Comment by encukou** [comment]
1-4. Thank you!
5. Thanks. I'm not an expert here -- I'm learning about ptys and their C API as I go. I'm happy that the test I added is hackable :)
6. The command is `make regen-configure`. If that doesn't work for you, run the `ubuntu:22.04` container (that is currently what the makefile uses, via `/Tools/build/regen-configure.sh`). If it's a problem, let me know and I'll regenerate this.  
    If you found `@tiran`'s outdated container mentioned in some *current* documentation, please let me know -- we should fix that. (Sadly we can't fix all the docs, like personal notes or shell histories...)
7. For the tests, local variable names don't matter much; it's OK to be a bit creative. For the API, this adds `fd` -- no problem there! For docs, it's *master*/*slave* for now; changing that will be a different discussion (in which the [Editorial board](https://peps.python.org/pep-0732/) might well decide to keep following POSIX).

Let's pretend `second_fd` stands for *fd of the second file*, and `reopened_second_fd` stands for *fd for the reopened second file*, or *reopened fd for the second file*. Does that work better? If not, change this to whatever feels best :)

**Comment by 8vasu** [comment]
@encukou you are very kind. Let me work on the reconfigure (and everything else you suggested).

**Comment by 8vasu** [comment]
@encukou Looks like all checks have passed! Your `/Tools/build/regen-configure.sh` suggestion worked like a charm... after I moved myself away from a directory that had colons (`:`) in its name as a time (hr, min, sec) separator; *docker* does not seem to like that.

File: Doc/library/os.rst
Comment (encukou) [inline Doc/library/os.rst:1450]: To avoid the docs warnings, C library functions need to go `nitpick_ignore` in `Doc/conf.py`.
Sorry for not catching this earlier. I pushed this directly to save a review round-trip.

**Review by encukou — APPROVED** [review]
+1 from me! 
@gpshead, do you want to take another look?

File: Modules/posixmodule.c
Comment (encukou) [reply Modules/posixmodule.c]: Filed here: #114727

**Comment by 8vasu** [comment]
@encukou Thank you for merging this! Looks like finally after years I have all the prerequisites for https://github.com/python/cpython/pull/101833. That one now needs to be updated and *maybe* broken up into a few parts in case it becomes too difficult to review: one for signal handling, one for winsize handling, one for adjusting the tests, etc. I will do this sometime during the next few months.

**Comment by encukou** [comment]
Thank you for perservering!
Looks like the main thing missing in 101833 is tests, so, a good strategy might be to send a PR whenever you get a test working :)

If I happen to miss a notification (= I don't reply for about a week), please ping me again.

**Comment by 8vasu** [comment]
@encukou

> Looks like the main thing missing in 101833 is tests

The tests for that were my first contribution to CPython in 2020. However, I still need to update the tests to use the various robust functions that I have added to `os`, `termios`, and `tty` over the years instead of the ad-hoc ones that I wrote in `Lib/test/test_pty.py`. If additional tests are required, I will definitely add them if you wish :)

**Pull request merged**
